File: filters/dolby.go
package filters

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/StollD/videoproc/common"
	"github.com/StollD/videoproc/mkv"
	"github.com/codeskyblue/go-sh"
	"github.com/natefinch/atomic"
	"github.com/ztrue/tracerr"
)

type DolbyAudioStream struct {
	mkv.AudioStream

	dir  string
	base mkv.AudioStream
}

func NormalizeDolby(base mkv.AudioStream, dir string) mkv.AudioStream {
	probe := base.Probe()

	val, exists := probe["codec_name"]
	if !exists {
		return base
	}

	if val.(string) != "ac3" {
		return base
	}

	return &DolbyAudioStream{base: base, dir: dir}
}

func (audio *DolbyAudioStream) Probe() map[string]interface{} {
	return audio.base.Probe()
}

func (audio *DolbyAudioStream) MediaInfo() map[string]interface{} {
	return audio.base.MediaInfo()
}

func (audio *DolbyAudioStream) ID() string {
	return audio.base.ID()
}

func (audio *DolbyAudioStream) Type() string {
	return audio.base.Type()
}

func (audio *DolbyAudioStream) Index() int {
	return 0
}

func (audio *DolbyAudioStream) Language() string {
	return audio.base.Language()
}

func (audio *DolbyAudioStream) Offset() float64 {
	return audio.base.Offset()
}

func (audio *DolbyAudioStream) Path() string {
	return filepath.Join(audio.dir, fmt.Sprintf("%s.norm.w64", audio.ID()))
}

func (audio *DolbyAudioStream) Prepare() error {
	return tracerr.Wrap(audio.base.Prepare())
}

func (audio *DolbyAudioStream) Process() error {
	// The file already exists
	if _, err := os.Stat(audio.Path()); !os.IsNotExist(err) {
		return nil
	}

	err := audio.base.Process()
	if err != nil {
		return tracerr.Wrap(err)
	}

	temp := common.Temp(audio.Path())

	args := make([]interface{}, 0)
	args = append(args, "-drc_scale", "0")

	info := audio.MediaInfo()

	val, exists := info["extra"]
	if exists {
		extra := val.(map[string]interface{})

		val, exists := extra["dialnorm_Average"]
		if exists {
			args = append(args, "-target_level", val.(string))
		}
	}

	args = append(args, "-i", audio.base.Path())
	args = append(args, "-map", fmt.Sprintf("0:%d", audio.base.Index()))
	args = append(args, "-codec", "pcm_f32le")
	args = append(args, "-y", temp)

	_, err = sh.Command(common.FFMPEG, args...).CombinedOutput()
	if err != nil {
		return tracerr.Wrap(err)
	}

	atomic.ReplaceFile(temp, audio.Path())

	// Cleanup
	err = audio.base.Cleanup()
	if err != nil {
		return tracerr.Wrap(err)
	}

	return nil
}

func (audio *DolbyAudioStream) Cleanup() error {
	err := os.Remove(audio.Path())
	if err != nil {
		return tracerr.Wrap(err)
	}

	return nil
}

func (audio *DolbyAudioStream) Samplerate() int {
	return audio.base.Samplerate()
}

File: src/filters/dolby.rs
use execute::Execute;
use std::{path::Path, process::Command};

use crate::{logging, mkv, utils::StrVec};

pub fn normalize(stream: &mkv::Stream, output: &Path) -> Result<mkv::Stream, ()> {
	logging::info!("Normalizing AC3 audio");

	let path = output.join(&stream.id).with_extension("norm.w64");

	let mut args = Vec::<String>::new();

	args.push_str("-drc_scale");
	args.push_str("0");

	if stream.dialnorm.is_some() {
		args.push_str("-target_level");
		args.push(format!("{}", stream.dialnorm.unwrap()));
	}

	args.push_str("-i");
	args.push_str(stream.path.to_str().unwrap());
	args.push_str("-map");
	args.push(format!("0:{}", stream.index));
	args.push_str("-codec");
	args.push_str("pcm_f32le");
	args.push_str("-y");
	args.push_str(path.to_str().unwrap());

	let cmd = Command::new("ffmpeg")
		.args(args)
		.execute_check_exit_status_code(0);

	if let Err(err) = cmd {
		logging::error!("Failed to run ffmpeg: {}", err);
		return Err(());
	}

	let mut new = stream.clone();
	new.path = path.clone();
	new.index = 0;
	new.codec = Some(String::from("pcm_f32le"));

	Ok(new)
}
